toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,409 @@
+		}
+		return err
+	})
+	if err != nil {
+		err = fmt.Errorf("failed to install '%s' in chroot: %w", dnfUtilsPackageName, err)
+		queryChroot.Close(false)
+		return
+	}
+	return
+}
+
+// getRepoPackages returns a list of packages available in the given repoUrl by running repoquery
+func getRepoPackages(repoUrl string) (packages []string, err error) {
+	const (
+		reqoqueryTool    = "repoquery"
+		randomNameLength = 10
+	)
+	var (
+		queryCommonArgList = []string{"-y", "-q", "--disablerepo=*", "-a", "--qf", "%{location}"}
+	)
+
+	logger.Log.Infof("Pre-cache: Getting package data from %s", repoUrl)
+
+	// We want to avoid using the same repo name for each repoUrl, so we generate a random name
+	randomName, err := randomization.RandomString(randomNameLength, randomization.LegalCharactersAlphaNum)
+	if err != nil {
+		err = fmt.Errorf("failed to generate random string: %w", err)
+		return
+	}
+	repoPathArg := fmt.Sprintf("--repofrompath=mariner-precache-%s,%s", randomName, repoUrl)
+	finalArgList := append(queryCommonArgList, repoPathArg)
+
+	onStdout := func(args ...interface{}) {
+		line := args[0].(string)
+		packages = append(packages, line)
+	}
+
+	// Run the repoquery command
+	err = shell.ExecuteLiveWithCallback(onStdout, logger.Log.Warn, true, reqoqueryTool, finalArgList...)
+	if err != nil {
+		err = fmt.Errorf("failed to run repoquery command: %w", err)
+		return
+	}
+
+	return
+}
+
+// downloadMissingPackages will attemp to download each package listed in rpmSnapshot that is not already present in the
+// outDir. It will return a list of the packages that were downloaded. It will use concurrentNetOps to limit the number of
+// concurrent network operations used to download the missing packages. It will also monitor the results and print periodic
+// progress updates to the console.
+func downloadMissingPackages(rpmSnapshot *repocloner.RepoContents, availablePackages map[string]string, outDir string, concurrentNetOps uint) (downloadedPackages []string, err error) {
+	timestamp.StartEvent("download missing packages", nil)
+	defer timestamp.StopEvent(nil)
+
+	// We will be downloading packages concurrently, so we need to keep track of when they are all done via a wait group. To
+	// simplify the code just use goroutines with a semaphore channel to limit the number of concurrent network operations. Each
+	// goroutine will handle a single package and will send the result of the download to a results channel. The main thread will
+	// monitor the results channel and update the progress counter accordingly.
+	wg := new(sync.WaitGroup)
+	netOpsSemaphore := make(chan struct{}, concurrentNetOps)
+	results := make(chan downloadResult)
+	doneChannel := make(chan struct{})
+
+	// Spawn a worker for each package, they will all do preliminary checks in parrallel before synchronizing on the semaphore.
+	// Each worker is responsible for adding and removing itself from the wait group.
```

Comment:
Nit-pick: the workers don't add themselves to wait groups, we're doing that still in the main thread.

Reply:
Stale comment, there was a race condition that necessitated this change. Fixed